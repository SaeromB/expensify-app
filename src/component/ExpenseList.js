import React from 'react';
// this connects react from redux store use at individual components 
import { connect } from 'react-redux';

const ExpenseList = (props) => (
  <div>
    <h1>ExpenseList</h1>
    {props.filters.text}
    {props.expenses.length}
  </div>
);

// what kind of info do we want from the store 
// state store gets passed in as a argument
 
// const ConnectedExpenseList = connect((state) => {
//   return {
//     expenses: state.expenses
//   };
// })(ExpenseList);

// When the store changes this will automatically rerun
// When the store changes the component will rerender the new values
const mapStateToProps = (state) => {
  return {
    expenses: state.expenses,
    filters: state.filters
  }
}
export default connect(mapStateToProps)(ExpenseList);