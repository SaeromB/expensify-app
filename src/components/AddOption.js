import React, {Component} from 'react';


// addOption 처럼 parent의 데이터를 바꿔야할 상황이 있다
class AddOption extends Component {
  constructor(props){
    super(props)
    this.handleAddOption = this.handleAddOption.bind(this)
    this.state = {
      error: undefined
    }
  }
  handleAddOption(e) {
    e.preventDefault();

    const option = e.target.elements.option.value.trim()
    const error = this.props.handleAddOption(option)

    // if (option) {
    //   this.props.handleAddOption(option)
    // }

    this.setState(() => ({error}));
  }
  render() {
    return(
      <div>
        {this.state.error && <p>{this.state.error}</p>}
        <form onSubmit={this.handleAddOption}>  
          <input type='text' name='option'/>
          <button>Add Option</button>
        </form>
      </div>
    )
  }
}

export default AddOption;

// stateless functional components can use 
// (props) is the same as this.props
// const User = (props) => {
//   return (
//     <div>
//       <p>Name: {props.name}</p>
//       <p>age: </p>
//     </div>
//   )
// }

