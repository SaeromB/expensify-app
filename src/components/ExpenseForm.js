import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
// import 'react-dates/initialize';
// import 'react-dates/lib/css/_datepicker.css';

// const date = new Date();

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props);

    this.state = {
      // if props.expese exist then show props.expense.description if not show empty string.
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      // we need the amount with a decimal and convert to String
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState (() => ({description}));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState (() => ({note}))
    // e.parsist()
    // this.setState({note: e.target.value})
  };

  // will return a date when it is clicked
  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(() => ({createdAt}))
    }
  };
  onFocusChange = ({focused}) => {
    this.setState(()=> ({calendarFocused : focused}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    // match(regex expressions)
    // if there is no amount or the amount is a match
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      this.setState(() => ({amount}))
    }
  }
  
  onSubmit = (e) => {
    e.preventDefault();
    
    if(!this.state.description || !this.state.amount){
      this.setState(() => ({error: 'Please provide description and amount'})) 
    } else {
      this.setState(()=>({error: ''}))
      // this props is passed down to AddExpensePage
      // Passed the data out of Expense form
      // props is passed in to the parent?
      // this makes use the form to be reused at AddExpensePage and EditExpensePage
      this.props.onSubmit({
        description: this.state.description,
        // amount is a string working at base 10
        amount: parseFloat(this.state.amount, 10 ) * 100,
        // this is a momnet object 
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      }) 
      // console.log('Submitted');
    }
  }
  
  render(){
    return( 
      <div>
        {this.state.error &&  <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}> 
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value = { this.state.description} onChange={this.onDescriptionChange} />
          <input
            type="text"
            placeholder="Amount"
            value = {this.state.amount} onChange={this.onAmountChange} />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            // this will make everysingle day avalible 
            isOutsideRange={()=> false }
            />
          <textarea
            placeholder=" Add a note for your expenses (optional)"
            value = {this.state.note} onChange={this.onNoteChange} >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}