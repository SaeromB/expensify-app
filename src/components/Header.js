import React from 'react';


const Header = (props) => (
  <div className="header">
    <h1 className="header__title">{props.title}</h1>
    {props.subtitle && <h2 className="header__subtitle">{props.subtitle}</h2>}
  </div>
)
  
  // give default to the title
Header.defaultProps = {
  title: 'Indecision'
}

export default Header;

