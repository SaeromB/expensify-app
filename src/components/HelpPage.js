import React from 'react';
import ReactDOM from 'react-dom';
import {BrowserRouter, Route, Switch, Link, NavLink} from 'react-router-dom'

const HelpPage = () => (
  <div>
    Help
  </div>
)

export default HelpPage;