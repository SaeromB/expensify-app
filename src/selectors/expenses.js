// Get visible expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    // if the typeof startDate is not number this will be filtered if not expense.createdAt(ex:1) will be checked if it is bigger then startDate
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    // if all of these are true and the item will be in the array if the result is false it will not show 
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if(sortBy === 'amount'){
      return a.amount < b.amount ? 1 : -1
    }
  });
};

export default getVisibleExpenses;