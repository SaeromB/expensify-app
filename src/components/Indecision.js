import React, {Component} from 'react';
import Header from './Header';
import Action from './Actions';
import Options from './Options';
import AddOption from './AddOption';
import OptionModal from './OptionModal';

class Indecision extends Component {

  state = {
    options: [],
    selectedOption: undefined
  }

  handleDeleteOptions = () => {
    // this.setState(() => {
    //   return {
    //   options: []
    //   }
    // })
    // If we want to return an object inside the function arrow is used as the function body so it need to be wrapped in ()
    this.setState (()=> ({options: []}));
  }

  //Remove indivisual item
  handleDeleteOption =(optionToRemove)=>{
    this.setState((prevState) => ({
    // update options array
    options : prevState.options.filter((option) => optionToRemove !== option)
    }))
  }
  
  handlePick = () => {
    const randomNum = Math.floor(Math.random() * this.state.options.length);
    const option = this.state.options[randomNum]
    alert(option)
  }

  handleAddOption = (option) => {
    if(!option) {
      return 'Enter valid value to add item'
    } else if (this.state.options.indexOf(option) > -1 ) {
      return 'This option exist'
    }
    // this.setState((prevState) => {
    //   return {
    //     options: prevState.options.concat(option)
    //   }
    this.setState ((prevState)=> ({ options: prevState.options.concat(option)}))
    // })
  }
  
  componentDidMount(){
    try {
      const json = localStorage.getItem('options')
      const options = JSON.parse(json);

      // this.setState(() => {
      //   return {
      //     options:options
      //   }
      // })
      if(options){
        this.setState(() => ({options}))
      }
    } catch(e){
      // do nothing
    }
  }

  //componentDidUpdate will work after props or state changes ex)change options
  componentDidUpdate(prevProps, prevState) {
    if(prevState.options.length !== this.state.options.length) {
      const json = JSON.stringify(this.state.options)
      localStorage.setItem('options', json);
    }
  }
  // if there is a web with different pages and render something new Unmount will work
  componentWillUnmount(){
    console.log('componentWillUnmount')
  }
  // We can delete the options by making a method and passing it down to a child component(<Option/>) and having it called down in the child component
  // that alow to reverse the data flow and the child component can do somthing in the parent component ex)handleDeleteOptions

  
  render(){
    const subtitle = 'Put your life in the hands of a Computer'
    return(
      <div>
        <Header subtitle={subtitle}/>
        {/* if there is options do not disable button */}
        <Action hasOptions={this.state.options.length > 0} handlePick={this.handlePick}/>
        {/* this will rerender (this is a prop)*/}
        <Options
          options={this.state.options}
          handleDeleteOptions={this.handleDeleteOptions}
          handleDeleteOption={this.handleDeleteOption} />
        <AddOption handleAddOption={this.handleAddOption} />
        <OptionModal />
      </div>
    )
  }
}

export default Indecision
