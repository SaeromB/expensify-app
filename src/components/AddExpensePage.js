// Going to use the BrowserRouter to create the Router once and Route for everysingle page
import React from 'react';
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {addExpense} from '../actions/expenses';

export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.onSubmit(expense);
    this.props.history.push('/')
  };
  render(){
    return (
      <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={this.onSubmit} />
      </div>  
    )
  }
}

// const AddExpensePage = (props) => (
//   <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm onSubmit={(expense)=> {
//       // Added the date to the redux-store
//       // props.dispatch(addExpense(expense));
//       props.onSubmit(expense);
//       // change pages
//       props.history.push('/');
//     }}/>
//   </div>
// );

// mapDispatchToProps a new wrapper component will subscribe to Redux store update
// when the store needs to updated mapDispatchToProps will be called

// const mapDispatchToProps = (dispatch) => {
//   return {
//     onSubmit: (expense) => dispatch(addExpense(expense))
//   }
// }

const mapDispatchToProps = (dispatch) => ({
  onSubmit : (expense) => dispatch(addExpense(expense))
});

// This needs to be connected in the store 
// Now props. dispatch can be used.
export default connect(undefined, mapDispatchToProps)(AddExpensePage);