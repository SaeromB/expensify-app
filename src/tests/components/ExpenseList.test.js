import React from 'react';
import {shallow} from 'enzyme';
// getting the unconnected version?
import { ExpenseList } from '../../components/ExpenseList';
import expenses from '../fixtures/expenses';
import '../setupTest';

test('Should render ExpenseList with expenses', () => {
  const wrapper = shallow(<ExpenseList expenses={expenses} />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseList with empty expense', () => {
  const wrapper = shallow(<ExpenseList expenses={[]} />);
  expect(wrapper).toMatchSnapshot();
})