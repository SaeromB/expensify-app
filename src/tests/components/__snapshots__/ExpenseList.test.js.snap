// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseList with empty expense 1`] = `ShallowWrapper {}`;

exports[`Should render ExpenseList with expenses 1`] = `ShallowWrapper {}`;
